### PR TITLE
Fix stack check overflows (primarily related to preemption)

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

File: src/objects.h [hunk 1]
```diff
@@ -680,12 +680,139 @@ class FrameCallback {
 };
 
 
+class Method {
```
Comment (kasperl) [inline src/objects.h:683]: Moved to make it usable from the Stack implementation.

File: src/interpreter_run.cc [hunk 2]
```diff
@@ -249,16 +249,27 @@ Interpreter::Result Interpreter::run() {
...
+    sp = load_stack(&pending);
+    static_assert(FRAME_SIZE == 2, "Unexpected frame size");
     Object* frame_marker = POP();
     ASSERT(frame_marker == program->frame_marker());
+    bcp = reinterpret_cast<uint8*>(POP());
+    // When we're preempted right when a stack needs to grow, we remember
+    // that we haven't done the correct stack overflow check and taken
+    // care of any necessary stack growth. We do that now that we're
+    // resuming interpretation. This is also another preemption check
+    // so we risk making no progress if we keep getting preempted.
```
Comment (floitsch) [inline src/interpreter_run.cc:268]: ?
```suggestion
    // When we are preempted at a call-site, we haven't done the
    // correct stack overflow check yet. We do the check now,
    // using the remembered 'pending' target method.
    // This is also another preemption check so we risk making no
    // progress if we keep getting preempted.
```

File: src/objects.h [hunk 3]
```diff
@@ -716,10 +850,30 @@ class Stack : public HeapObject {
...
+  static const int GUARD_ZONE_WORDS = 4;
+#endif
+  static const int GUARD_ZONE_SIZE = GUARD_ZONE_WORDS * WORD_SIZE;
+
   static const int LENGTH_OFFSET = HeapObject::SIZE + WORD_SIZE;
   static const int TOP_OFFSET = LENGTH_OFFSET + WORD_SIZE;
   static const int TRY_TOP_OFFSET = TOP_OFFSET + WORD_SIZE;
-  static const int HEADER_SIZE = TRY_TOP_OFFSET + WORD_SIZE;
+  static const int PENDING_STACK_CHECK_METHOD_OFFSET = TRY_TOP_OFFSET + WORD_SIZE;
+  static const int GUARD_ZONE_OFFSET = PENDING_STACK_CHECK_METHOD_OFFSET + WORD_SIZE;
```
Comment (floitsch) [inline src/objects.h:875]: Add a comment that the stack grows down.

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/objects.h (hunk 3 above)
Comment (kasperl) [reply src/objects.h:875]: Done.

**Pull request merged**